### PR TITLE
MangAdventure split

### PR DESCRIPTION
**Pull request by ObserverOfTime**
Do not merge until the source generator supports URL intents.

They can be implemented as a list of `Triple(host, pathPattern, scheme)` in `ThemeSourceData` so the generator can add them to the manifest. `ThemeSourceData` will also need another field that declares the name of the activity class.

Also, it would be nice if the generator could replace hyphens and other invalid identifier characters in the name.

### CODE REVIEW

**Comment by arkon** [comment]
Were you planning on making the change to the generator, or hoping that @AriaMoradi does it?

**Comment by ObserverOfTime** [comment]
I can do it next weekend if no one else does it by then.

**Comment by AriaMoradi** [comment]
> They can be implemented as a list of `Triple(host, pathPattern, scheme)` in `ThemeSourceData` so the generator can add them to the manifest. `ThemeSourceData` will also need another field that declares the name of the activity class.

This multi-source is maybe the only one that needs this kind of feature so I'm against adding it to the library. 

> Also, it would be nice if the generator could replace hyphens and other invalid identifier characters in the name.

The Source name can be any string of Unicode characters,  and adding this feature doesn't seem wise to me, you can already provide pakage name and file(class) name overrides manually. 
Also src override gives you all the power you need to pull off this split. 
The `ThemeSource` just acts as a definition tool for sources which essentially replaces the factory class and it does a minimal source generation for sources that don't need overrides.

**Comment by arkon** [comment]
> This multi-source is maybe the only one that needs this kind of feature so I'm against adding it to the library.

Are you saying this can be achieved with the overrides rather than having to modify the generator code, or are you against multi-source modules having the ability to handle intents?

**Comment by AriaMoradi** [comment]
> > This multi-source is maybe the only one that needs this kind of feature so I'm against adding it to the library.
> 
> Are you saying this can be achieved with the overrides rather than having to modify the generator code, or are you against multi-source modules having the ability to handle intents?

Any shared source file should be put in the `multisrc.multisrcname` package and any source-specific code should be in the src override.
If you take a look at the logic, it copies any class file that doesn't end with `Generator.kt` (`ThemeSourceGenerator.kt` line 114)

**Comment by arkon** [comment]
Sure, but it doesn't handle overriding the manifest, does it?

**Comment by AriaMoradi** [comment]
> Sure, but it doesn't handle overriding the manifest, does it?

Ah, I completely forgot that manifest needs overrides to handle intents, oops.
we can add a `manifest` directory to `overrides` and handle it that way or put it in the src override directory and have the generator look for a manifest file?

**Comment by ObserverOfTime** [comment]
That should work but you also need to handle the activity class.

**Comment by AriaMoradi** [comment]
> That should work but you also need to handle the activity class.

Put inside `multisource.themename`? what's special about he activity class?

**Comment by ObserverOfTime** [comment]
It just accepts the intent and sends it to the main app.

**Comment by AriaMoradi** [comment]
> It just accepts the intent and sends it to the main app but it needs to be declared in `build.gradle`.

Noting is delared in https://github.com/tachiyomiorg/tachiyomi-extensions/blob/81f9eede50758261f46b888a518c126221860d74/src/all/mangadventure/build.gradle ?

**Comment by ObserverOfTime** [comment]
Oh, right, it's in the manifest.

**Comment by ObserverOfTime** [comment]
It should be ready to merge now.

File: multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangadventure/MangAdventureGenerator.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+/** [MangAdventure] source generator. */
+class MangAdventureGenerator : ThemeSourceGenerator {
+    override val themePkg = "mangadventure"
+
+    override val themeClass = "MangAdventure"
+
+    override val baseVersionCode = 1
+
+    override val sources = listOf(
+        SingleLang("Arc-Relight", "https://arc-relight.com", "en", className = "ArcRelight"),
```
Comment (AriaMoradi) [inline multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangadventure/MangAdventureGenerator.kt:15]: Only one source?!
Why are you doing multisrc then?!

**Review by AriaMoradi — CHANGES_REQUESTED** [review]

File: multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangadventure/MangAdventureGenerator.kt (hunk 1 above)
Comment (ObserverOfTime) [reply multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangadventure/MangAdventureGenerator.kt:15]: Only one source *now*. There was another one that changed to Madara and there may be more in the future.

**Pull request merged**